ding/policy/fqf.py

```diff
@@ -0,0 +1,294 @@
+
+        # ====================
+        # Q-learning update
+        # ====================
+        self._quantile_loss_optimizer.zero_grad()
+        quantile_loss.backward()
+        if self._cfg.learn.multi_gpu:
+            self.sync_gradients(self._learn_model)
+        total_norm_Q = torch.norm(
+            torch.stack([torch.norm(p.grad.detach(), 2.0) for p in self._model.head.Q.parameters()]), 2.0
+        )
+        total_norm_fqf_fc = torch.norm(
+            torch.stack([torch.norm(p.grad.detach(), 2.0) for p in self._model.head.fqf_fc.parameters()]),
+            2.0
+        )
+        total_norm_encoder = torch.norm(
+            torch.stack([torch.norm(p.grad.detach(), 2.0) for p in self._model.encoder.parameters()]), 2.0
+        )
+        self._quantile_loss_optimizer.step()
+
+        # =============
+        # after update
+        # =============
+        self._target_model.update(self._learn_model.state_dict())
+        return {
+            'cur_lr_fraction_loss': self._fraction_loss_optimizer.defaults['lr'],
+            'cur_lr_quantile_loss': self._quantile_loss_optimizer.defaults['lr'],
+            'logit': logit.mean().item(),
+            'fraction_loss': fraction_loss.item(),
+            'quantile_loss': quantile_loss.item(),
+            'total_norm_quantiles_proposal': total_norm_quantiles_proposal,
+            'total_norm_Q': total_norm_Q,
+            'total_norm_fqf_fc': total_norm_fqf_fc,
+            'total_norm_encoder': total_norm_encoder,
+            'priority': td_error_per_sample.abs().tolist(),
+            # Only discrete action satisfying len(data['action'])==1 can return this and draw histogram on tensorboard.
+            '[histogram]action_distribution': data['action'],
+            '[histogram]quantiles_hats': quantiles_hats[0],
+        }
+
+    def _monitor_vars_learn(self) -> List[str]:
+        return [
+            'cur_lr_fraction_loss', 'cur_lr_quantile_loss', 'logit', 'fraction_loss', 'quantile_loss',
+            'total_norm_quantiles_proposal', 'total_norm_Q', 'total_norm_fqf_fc', 'total_norm_encoder'
+        ]
+
+    def _state_dict_learn(self) -> Dict[str, Any]:
+        return {
+            'model': self._learn_model.state_dict(),
+            'target_model': self._target_model.state_dict(),
+            'optimizer_fraction_loss': self._fraction_loss_optimizer.state_dict(),
+            'optimizer_quantile_loss': self._quantile_loss_optimizer.state_dict(),
+        }
+
+    def _load_state_dict_learn(self, state_dict: Dict[str, Any]) -> None:
+        self._learn_model.load_state_dict(state_dict['model'])
+        self._target_model.load_state_dict(state_dict['target_model'])
+        #self._optimizer.load_state_dict(state_dict['optimizer'])
+        self._fraction_loss_optimizer.load_state_dict(state_dict['optimizer_fraction_loss'])
+        self._quantile_loss_optimizer.load_state_dict(state_dict['optimizer_quantile_loss'])
+
+    def _init_collect(self) -> None:
```

Comment:
don't overwrite the same methods as DQNPolicy, such as `_init_collect`, `_get_train_sample`